src/lib/crypto/kimchi_backend/common/plonk_constraint_system.ml

```diff
@@ -922,8 +930,54 @@ end = struct
                 let value = compute lc in
                 res.(col_idx).(row_idx) <- value ;
                 Hashtbl.set internal_values ~key:var ~data:value ) ) ;
+
+    let module MapRuntimeTable = struct
+      module T = struct
+        type t = int32 * Fp.t [@@deriving hash, sexp, compare]
+      end
+
+      include T
+      include Core_kernel.Hashable.Make (T)
+    end in
+    let map_runtime_tables = MapRuntimeTable.Table.create () in
+    let runtime_tables : Fp.t Kimchi_types.runtime_table array =
+      match sys.runtime_tables_cfg with
+      | Unfinalized_runtime_tables_cfg_rev _ ->
+          failwith
+            "Attempted to generate a witness for an unfinalized constraint \
+             system"
+      | Compiled_runtime_tables_cfg cfgs ->
+          Array.mapi cfgs ~f:(fun rt_idx { Kimchi_types.id; first_column } ->
+              let data =
+                Array.mapi first_column ~f:(fun i v ->
+                    ignore
+                    @@ MapRuntimeTable.Table.add map_runtime_tables ~key:(id, v)
+                         ~data:(i, rt_idx) ;
+                    (* default padding value for lookup *)
+                    Fp.zero )
+              in
+              let rt : Fp.t Kimchi_types.runtime_table = { id; data } in
+              rt )
+    in
+
+    (* Fill in the used entries of the runtime lookup tables. *)
+    List.iter sys.runtime_lookups ~f:(fun (id, (idx, v)) ->
+        let compute_value x = compute ([ (Fp.one, x) ], None) in
+        let vid = compute_value id in
+        let vidx = compute_value idx in
+        let vv = compute_value v in
+        (* FIXME: we should have a int32 here. We are not sure the ID will be a
+           int32. We should enforce that.
+           See https://github.com/MinaProtocol/mina/issues/13955
+        *)
+        let id_int32 = Int32.of_string @@ Fp.to_string vid in
```

Comment:
Wow. We don't have a `Fp.to_int32` function? I'm sad.

Reply:
It makes sense to not have it as it is not safe. I would be against having it. If we add one, we should call it `Fp.to_int32_opt/exn` or `Fp.unsafe_to_int32`.

Reply:
Got you. But I don't see how  making the conversion using strings makes anything safer.

Reply:
It does not. I would simply not expose a function in the interface, or, if we do, explicity mention the unsafe part in the name.

Reply:
I opened https://github.com/MinaProtocol/mina/issues/13992 to keep track of it. Closing the thread.